---
title: "Python 小手册"
subtitle: "对象析构 | 字符处理 "
layout: post
author: "luoruiqing"
header-style: text
catalog:    true
tags:
  - Python
---



## 数据操作


#### 缓存属性

```py
try:
    from functools import cached_property  # py38+
except ImportError:
    from django.utils.functional import cached_property # 需要安装django


@cached_property
def a():
    return 1
```

#### 定长列表

```py
from collections import deque # 定长列表

```

## 文件类


#### 清除文件

```python
from glob import glob
import os

items = ["*.flv", "*.mp4", "*.f4v", "../*.flv", "../*.mp4", "../*.pyc", "../*.download"]
for item in items:
    for file in glob(item):
        if os.path.exists(file):
            os.remove(file)
```


## 定制

#### 错误定制

```py
class BaseError(BaseException):
    """ 自定义错误基类 """
    CODES = {}  # 所有错误


class Error(BaseError):
    ''' 自定义错误类 '''
    status_code = 400  # 默认错误响应状态码
    code = None  # 错误代码
    message = ''  # 错误消息

    def __new__(cls, code, *args, **kwargs):
        check = kwargs.pop('check', True)
        instance = super().__new__(cls, code, *args, **kwargs)
        if check:
            assert code not in instance.CODES, f"Error code duplication: {code}."
            instance.CODES[code] = instance  # 记录错误实例
        return instance

    def __init__(self, code, message, status_code=400, *args, **kwargs):
        ''' 带参数进入格式化到消息体 '''
        self.code = code
        self.message = message
        self.status_code = status_code

    def __call__(self, *args, **kwargs):
        ''' 回调始终生成新的错误对象, 且code不允许变化 '''
        message = (self.message.format(*args, **kwargs)) if args or kwargs else self.message

        return self.__class__(
            code=self.code,
            message=message,
            status_code=self.status_code,
            check=False,
        )

    def __str__(self):
        return f'Error {self.code}({self.status_code}): {self.message}.'


if __name__ == "__main__":
    import unittest

    class TestError(unittest.TestCase):
        ''' 单元测试 '''

        def tearDown(self):
            Error.CODES.clear()  # 清除

        def test_error_code_duplication(self):
            ''' 测试错误代码重复 '''
            ERROR1 = Error(10000, '这是个错误! {}')
            with self.assertRaises(AssertionError):
                ERROR2 = Error(10000, '这是个错误! {}')

        def test_error_repeat_call(self):
            ''' 测试多次调用同一实例 '''
            ERROR1 = Error(10000, '这是个错误! {}')
            temporary_error_1 = ERROR1('附加信息1')
            self.assertEqual(temporary_error_1.code, 10000)
            self.assertEqual(temporary_error_1.message, '这是个错误! 附加信息1')
            temporary_error_2 = temporary_error_1('附加信息2')
            self.assertEqual(temporary_error_2.code, 10000)
            self.assertEqual(temporary_error_2.message, '这是个错误! 附加信息1')

        def test_error_type_checking(self):
            ''' 测试实例类型 '''
            ERROR1 = Error(10000, '这是个错误! {}')
            self.assertIsInstance(ERROR1, Error)
            temporary_error_1 = ERROR1('附加信息1')
            self.assertIsInstance(temporary_error_1, Error)
            temporary_error_2 = temporary_error_1('附加信息2')
            self.assertIsInstance(temporary_error_2, Error)

            self.assertTrue(issubclass(ERROR1.__class__, BaseError))

    unittest.main()
```


#### 错误位置



#### 获取所有类